### PR TITLE
 refact: return detailed error messages for WebApi integration

### PR DESCRIPTION
**Pull request by omsmith**
Closes: https://github.com/Brightspace/D2L.Security.OAuth2/issues/40

### CODE REVIEW

File: test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs [hunk 1]
```diff
@@ -29,14 +29,24 @@ await TestUtilities
...
-		public async Task ImpersonationToken_Forbidden() {
+		public async Task ImpersonationToken_Unauthorized() {
 			string jwt = await TestUtilities
 				.GetAccessTokenValidForAMinute( userId: 1234, actualUserId: 1235 )
 				.SafeAsync();
 
-			await TestUtilities
-				.RunBasicAuthTest( "/authorization/imp", jwt, HttpStatusCode.Forbidden )
+			var response = await TestUtilities
+				.RunBasicAuthTest<OAuth2ErrorResponse>( "/authorization/imp", jwt, HttpStatusCode.Unauthorized )
```
Comment (omsmith) [inline test/D2L.Security.OAuth2.WebApi.IntegrationTests/Authorization/NoImpersonationAttributeTests.cs]: I changed these cause of the RFC suggestion, and I didn't think it super terribly mattered.

```
invalid_token
         The access token provided is expired, revoked, malformed, or
         invalid for other reasons.  The resource SHOULD respond with
         the HTTP 401 (Unauthorized) status code.  The client MAY
         request a new access token and retry the protected resource
         request.
```

File: src/D2L.Security.OAuth2.WebApi/Authorization/AuthenticationAttribute.cs [hunk 2]
```diff
@@ -46,10 +49,24 @@ public sealed class AuthenticationAttribute : OAuth2AuthorizeAttribute {
 					return false;
 
 				case PrincipalType.User:
-					return m_allowUsers;
+					if( m_allowUsers ) {
+						return true;
```
Comment (j3parker) [inline src/D2L.Security.OAuth2.WebApi/Authorization/AuthenticationAttribute.cs:53]: test comment 1

Comment (j3parker) [inline src/D2L.Security.OAuth2.WebApi/Authorization/AuthenticationAttribute.cs:53]: test comment 1

File: src/D2L.Security.OAuth2.WebApi/Authorization/AuthenticationAttribute.cs [hunk 3]
```diff
@@ -46,10 +49,24 @@ public sealed class AuthenticationAttribute : OAuth2AuthorizeAttribute {
...
+
+					throw new OAuth2Exception(
+						error: OAuth2Exception.Type.invalid_token,
+						errorDescription: "Users are not allowed to access this API."
+					);
 
 				case PrincipalType.Service:
-					return m_allowServices;
+					if( m_allowServices ) {
+						return true;
```
Comment (j3parker) [inline src/D2L.Security.OAuth2.WebApi/Authorization/AuthenticationAttribute.cs:63]: test comment 2

**Comment by j3parker** [comment]
We're going to deprecate the WebAPI library because it is incompatible with ASP.NET Core.

**Pull request closed without merging**